cluster-autoscaler/expander/grpcplugin/grpc_client.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpcplugin
+
+import (
+	"context"
+	"log"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/expander"
+	"k8s.io/autoscaler/cluster-autoscaler/expander/grpcplugin/protos"
+	"k8s.io/klog/v2"
+	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type grpcclientstrategy struct {
+	grpcClient protos.ExpanderClient
+}
+
+// NewFilter returns an expansion filter that creates a gRPC client, and calls out to a gRPC server
+func NewFilter(expanderCert string, expanderUrl string) expander.Filter {
+	client := createGRPCClient(expanderCert, expanderUrl)
+	if client == nil {
+		return &grpcclientstrategy{grpcClient: nil}
+	}
+	return &grpcclientstrategy{grpcClient: client}
+}
+
+func createGRPCClient(expanderCert string, expanderUrl string) protos.ExpanderClient {
+	var dialOpt grpc.DialOption
+
+	// if no Cert file specified, use insecure
+	if expanderCert == "" {
+		dialOpt = grpc.WithInsecure()
+	} else {
+		creds, err := credentials.NewClientTLSFromFile(expanderCert, "")
+		if err != nil {
+			log.Fatalf("Failed to create TLS credentials %v", err)
+			return nil
+		}
+		dialOpt = grpc.WithTransportCredentials(creds)
+	}
+	klog.V(2).Info("Dialing ", expanderUrl, " dialopt: ", dialOpt)
+	conn, err := grpc.Dial(expanderUrl, dialOpt)
+	if err != nil {
+		log.Fatalf("fail to dial server: %v", err)
+		return nil
+	}
+	return protos.NewExpanderClient(conn)
+}
+
+func (g *grpcclientstrategy) BestOptions(expansionOptions []expander.Option, nodeInfo map[string]*schedulerframework.NodeInfo) []expander.Option {
+	if g.grpcClient == nil {
+		log.Fatalf("Incorrect gRPC client config, filtering no options")
+		return expansionOptions
+	}
+
+	// Transform inputs to gRPC inputs
+	nodeGroupIDOptionMap := make(map[string]expander.Option)
+	grpcOptionsSlice := []*protos.Option{}
+	populateOptionsForGRPC(expansionOptions, nodeGroupIDOptionMap, &grpcOptionsSlice)
```

Comment:
It seems much more natural to create datastructures in the function and return them rather than mutate them this way. Is there any particular reason for the current approach?